libevmasm/Inliner.cpp

```diff
@@ -196,19 +194,38 @@ bool Inliner::shouldInlineFullFunctionBody(size_t _tag, ranges::span<AssemblyIte
 	return false;
 }
 
-
-optional<AssemblyItem::JumpType> Inliner::shouldInline(size_t _tag, AssemblyItem const& _jump, InlinableBlock const& _block) const
+optional<AssemblyItem> Inliner::shouldInline(size_t _tag, AssemblyItem const& _jump, InlinableBlock const& _block) const
 {
-	AssemblyItem exitJump = _block.items.back();
-	assertThrow(_jump == Instruction::JUMP && exitJump == Instruction::JUMP, OptimizerException, "");
+	assertThrow(_jump == Instruction::JUMP, OptimizerException, "");
+	AssemblyItem blockExit = _block.items.back();
 
 	if (
 		_jump.getJumpType() == AssemblyItem::JumpType::IntoFunction &&
-		exitJump.getJumpType() == AssemblyItem::JumpType::OutOfFunction
+		blockExit == Instruction::JUMP &&
+		blockExit.getJumpType() == AssemblyItem::JumpType::OutOfFunction &&
+		shouldInlineFullFunctionBody(_tag, _block.items, _block.pushTagCount)
+	)
+	{
+		blockExit.setJumpType(AssemblyItem::JumpType::Ordinary);
+		return blockExit;
+	}
+
+	// Inline small blocks, if the jump to it is ordinary or the blockExit is a terminating instruction.
+	if (
+		_jump.getJumpType() == AssemblyItem::JumpType::Ordinary ||
+		SemanticInformation::terminatesControlFlow(blockExit)
 	)
-		return
-			shouldInlineFullFunctionBody(_tag, _block.items, _block.pushTagCount) ?
-			make_optional(AssemblyItem::JumpType::Ordinary) : nullopt;
+	{
+		static AssemblyItems const jumpPattern = {
+			AssemblyItem{PushTag},
+			AssemblyItem{Instruction::JUMP},
+		};
+		if (
+			GasMeter::dataGas(codeSize(_block.items), m_isCreation, m_evmVersion) <=
+			GasMeter::dataGas(codeSize(jumpPattern), m_isCreation, m_evmVersion)
+		)
```

Comment:
This can of course be extended to something that considers execution cost as well. But if the data cost is smaller already, then it's definitely good to inline.

Reply:
I did quickly try some adapted version of the full-function-inlining-heuristics here - and there are quite some gains to be achieved by extending this - but I must have been missing something, because both deploy cost and runtime cost that involves contract creation tended to get significantly more expensive (while pure runtime savings on the other hand were quite nice)... so this will need some more tweaking. The basic version should in this PR should definitely be fine in any case, though.

Reply:
Let's merge it for now, we can still tweak it.